### DISCUSSION TITLE
InvalidOperationException for OpenGeneric, No Setter and IgnoreAllPropertiesWithAnInaccessibleSetter

### DISCUSSION THREAD

**Discussion by JoesGab** [LuckyPennySoftware/AutoMapper#4158] [NA]
### Source/destination types

```csharp
        public class A
        {
            public string some { get; set; }
        }

        public class B<T>
        {
            public string some { get; }
        }
```

### Mapping configuration

```csharp
                cfg.CreateMap(typeof(A), typeof(B<>))
                    .IgnoreAllPropertiesWithAnInaccessibleSetter();
```

### Version: x.y.z
12.0.1-preview.0.3
12.0.0
11.0.0

### Expected behavior

Map should work

### Actual behavior

Map throws System.InvalidOperationException: Sequence contains no matching element

### Steps to reproduce

```csharp
 var config = new MapperConfiguration(cfg =>
            {
                cfg.CreateMap(typeof(A), typeof(B<>))
                    .IgnoreAllPropertiesWithAnInaccessibleSetter();
            });

            config.AssertConfigurationIsValid();
```

Note that either adding a setter or removing the generic type "fixes" the Exception

**Comment by lbargaoanu** [comment]
You get an exception because you ignore smth that's ignored by default Not a big deal if you ask me. But you're welcome to fix it if it bothers you.

**Reply by JoesGab** [reply]
Well, I just boiled it down to a minimal sample. If you put that IgnoreAllPropertiesWithAnInaccessibleSetter in a more global fashion (e.g. ForAllMaps), then a situation like that will be an issue.

If you have any clue on where to start looking, I'd gladly try and fix.

**Reply by lbargaoanu** [reply]
Maybe start with the exception? I guess what you want is to ignore the ignore :)

**Reply by lbargaoanu** [reply]
But maybe `ShouldMapProperty` is a better way to achieve what you need. Or `ForAllPropertyMaps`.

**Reply by JoesGab** [reply]
So... if AutoMapper does not initially map properties without a setter by design, IgnoreAllPropertiesWithAnInaccessibleSetter should not even attempt to add an Ignore in such a case. Correct?

**Reply by lbargaoanu** [reply]
No, if you want to go in that direction, you can do that in your own code. Here you have to handle the general case, any `ForMember` seems to fail for such members in open generics.

**Reply by JoesGab** [reply]
I tracked my issue down to a change in commit e9702601, where the `Configure` for the `MemberConfigurationExpression` fails for a destination member with generic parameters.

[Line 133 here](https://github.com/AutoMapper/AutoMapper/blob/73ba4003e49e37e5fffec49b67e7d9ea1cbb8a8a/src/AutoMapper/Configuration/MemberConfigurationExpression.cs#L133) used to be reliant on Get-Accessors, not Write-Accessors.
This could be reverted with some additions in `TypeMap`, but I would still argue that automated adding of an Ignore for the property `some` should not be done to begin with in this case.

**Reply by lbargaoanu** [reply]
That revert doesn't make sense, only write accessors can be mapped.

**Reply by JoesGab** [reply]
As I see it, there's two cases.

For one, like you mentioned, any ForMember for a destination property without a setter will fail (no matter the options used). E.g. 
```csharp
cfg.CreateMap(typeof(A), typeof(B<>))
   .ForMember("some", opt => opt.Ignore());
```
Configuring those properties does not make much sense, so I would argue that a proper Exception should be thrown in this case. After all, this would already be done for misspelled properties too.

The other option would really only be to completely ignore the Configuration, which sounds very odd to me. 

The other thing is the IgnoreAllPropertiesWithAnInaccessibleSetter, which also leads to automated configuration of properties without a setter. In this case, I would think that PropertiesWithAnInaccessibleSetter should not return those properties, so no configuration will take place.

Would you agree?

**Reply by lbargaoanu** [reply]
No, for better or worse, we have to do what the non generic maps do. I didn't look deeply into it, but it seems they go about their business as nothing happened.

**Reply by JoesGab** [reply]
OK, I just find this design somewhat odd because it silently ignores what the user puts for a destination with no setter, e.g. this is perfectly valid:

```csharp
namespace AutoMapper.UnitTests
{
    public class IgnoreAllPropertiesWithAnInaccessibleSetterTests
    {
        public class Source
        {
            public int MapFromHere { get; set; }
        }

        public class Destination
        {
            public int MapToHere { get;  }
        }

        [Fact]
        public void AutoMapper_ShouldNotMapPropertyWithoutSet()
        {
            // Arrange
            var config = new MapperConfiguration(cfg =>
            {
                cfg.CreateMap<Source, Destination>()
                    .ForMember(dest => dest.MapToHere, opt => opt.MapFrom(src => src.MapFromHere)); // This map will be ignored.
            });
            var mapper = config.CreateMapper();

            var source = new Source { MapFromHere = 666, };

            // Act
            var destination = mapper.Map<Destination>(source);

            // Assert
            config.AssertConfigurationIsValid();
            Assert.Equal(0, destination.MapToHere);
        }
    }
}
```

If this is the intended way though, then I'm back to changing line 133 as suggested before. Mapping a read-only accessor indeed does not make sense, but there is otherwise no way I can think of for configuring the read-only property as it does right now.

**Reply by lbargaoanu** [reply]
[I was wrong](https://github.com/AutoMapper/AutoMapper/pull/4097/commits/d469fdf1212e7bb758d592c04c967c56b10dc0b4). Try the [MyGet](https://docs.automapper.org/en/latest/The-MyGet-build.html) build. For the cases where that config is useless, you can write [your own validations](https://docs.automapper.org/en/latest/Configuration-validation.html#custom-validations).

**Reply by JoesGab** [reply]
Works on the mapping_inheritance branch, for some reason not with the MyGet build.

**Reply by lbargaoanu** [reply]
You need to clear the NuGet cache or remove the package by hand.

**Reply by JoesGab** [reply]
🤦 Oh well, now I get it. 😄 Thank you! Works as expected now.
